src/main/java/org/opensearch/knn/search/NativeEngineKnnVectorQuery.java

```diff
@@ -0,0 +1,282 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+@RequiredArgsConstructor
+public class NativeEngineKnnVectorQuery extends Query {
+
+    private final KNNQuery knnQuery;
+
+    public Query rewrite(final IndexSearcher indexSearcher) throws IOException {
+        final IndexReader reader = indexSearcher.getIndexReader();
+        final KNNWeight knnWeight = (KNNWeight) knnQuery.createWeight(indexSearcher, ScoreMode.COMPLETE, 1);
+        List<LeafReaderContext> leafReaderContexts = reader.leaves();
+
+        List<Callable<TopDocs>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> searchLeaf(leafReaderContext, knnWeight));
+        }
+        TopDocs[] perLeafResults = indexSearcher.getTaskExecutor().invokeAll(tasks).toArray(TopDocs[]::new);
+        TopDocs topK = TopDocs.merge(knnQuery.getK(), perLeafResults);
+        if (topK.scoreDocs.length == 0) {
+            return new MatchNoDocsQuery();
+        }
+        return createRewrittenQuery(reader, topK);
+    }
+
+    private Query createRewrittenQuery(IndexReader reader, TopDocs topK) {
+        int len = topK.scoreDocs.length;
+        Arrays.sort(topK.scoreDocs, Comparator.comparingInt(a -> a.doc));
+        int[] docs = new int[len];
+        float[] scores = new float[len];
+        for (int i = 0; i < len; i++) {
+            docs[i] = topK.scoreDocs[i].doc;
+            scores[i] = topK.scoreDocs[i].score;
+        }
+        int[] segmentStarts = findSegmentStarts(reader, docs);
+        return new DocAndScoreQuery(knnQuery.getK(), docs, scores, segmentStarts, reader.getContext().id());
+    }
+
+    static int[] findSegmentStarts(IndexReader reader, int[] docs) {
+        int[] starts = new int[reader.leaves().size() + 1];
+        starts[starts.length - 1] = docs.length;
+        if (starts.length == 2) {
+            return starts;
+        }
+        int resultIndex = 0;
+        for (int i = 1; i < starts.length - 1; i++) {
+            int upper = reader.leaves().get(i).docBase;
+            resultIndex = Arrays.binarySearch(docs, resultIndex, docs.length, upper);
+            if (resultIndex < 0) {
+                resultIndex = -1 - resultIndex;
+            }
+            starts[i] = resultIndex;
+        }
+        return starts;
+    }
+
+    private TopDocs searchLeaf(LeafReaderContext ctx, KNNWeight queryWeight) throws IOException {
+        int totalHits = 0;
+        KNNScorer scorer = (KNNScorer) queryWeight.scorer(ctx);
+        DocIdSetIterator docIdSetIterator = scorer.iterator();
+        List<ScoreDoc> scoreDocs = new ArrayList<>();
+        while (docIdSetIterator.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+            scoreDocs.add(new ScoreDoc(docIdSetIterator.docID(), scorer.score()));
+            totalHits++;
+        }
```

Comment:
I would really love to see how we can avoid this iteration. may be instead of calling `queryWeight.scorer` we can add another function in KNNWeightClass which can by `queryWeight.scorer` and searchLeaf function to get docs and Score in an easy format.

Reply:
Took a deeper look into this, I initially thought that DISI is sorting scores with the BulkAdder. But its sorting Ids and not scores. I will have to revisit this since `TopDocs.merge` expects sorted TopDocs by score.